src/transformers/trainer.py

```diff
@@ -342,6 +393,131 @@ def __init__(
         self.label_names = default_label_names if self.args.label_names is None else self.args.label_names
         self.control = self.callback_handler.on_init_end(self.args, self.state, self.control)
 
+    def get_num_training_steps(self):
+        """
+        number of training steps is either 1. args.max_steps if --max_steps > 1 2. else derive from dataset if we can
+        get its size
+        """
+        if self.args.max_steps > 0:
+            return self.args.max_steps
+        elif isinstance(self.train_dataset, collections.abc.Sized):
+            # XXX: hack to get dataset len before dist world is initiated
+            train_dataloader = DataLoader(
+                self.train_dataset,
+                batch_size=self.args.train_batch_size,
+                collate_fn=self.data_collator,
+                drop_last=self.args.dataloader_drop_last,
+            )
+            num_update_steps_per_epoch = len(train_dataloader) // self.args.gradient_accumulation_steps
+            num_update_steps_per_epoch = max(num_update_steps_per_epoch, 1)
+            return math.ceil(self.args.num_train_epochs * num_update_steps_per_epoch)
```

Comment:
This will not take into account the number of GPUs in a distributed training. Does DeepSpeed correct internally for that? If that's the case, maybe we could just use the math behind `DataLoader.__len__` and not instantiate a full `DataLoader` for nothing?

Reply:
I haven't followed the code, but my intuition is telling me that ds should sort it out and require the total - otherwise how a user is to figure that out as they switch between different number of gpus - that would be too error-prone.

But I will ask to be sure.

Reply:
So you're suggesting to copy this exact part:

https://github.com/pytorch/pytorch/blob/c439a6534de80be5ef30bfb39633a83cdba74838/torch/utils/data/dataloader.py#L395-L402

and probably assert if it's a custom sampler? Not sure what can we do if the latter is used?

Reply:
Just the part
```
if self.drop_last:
    length = length // self.batch_size
else:
    length = ceil(length / self.batch_size)
```
Trainer does not support custom sampler anyway, so I think we should focus on the simple case.